doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,592 @@
+- ROWS: similar to GROUPS, but use indexes that [include the hash](#Peer-groups).
+
+There is also `frame_exclusion`, which sometimes necessitates special handling for the group that contains the current row. In such cases, we will put together the result value of the window function from 3 parts: 1. prefix sum (generalized to arbitrary intervals) for groups that are earlier than the current row’s group, 2. prefix sum for groups that are later than the current row’s group, 3. current row’s group (without prefix sum).
+
+*Solving II.:*
+
+For invertible aggregation functions (e.g., sum, but not min/max), it would be tempting to use the existing prefix sum with a minor trick: `agg(a,b) = agg(0,b) - agg(0,a)`. However, the performance of this wouldn't be good, because even if `(a,b)` is a small interval, the `(0,a)` and the `(0,b)` intervals will be big, so there will be many changes of the aggregates of these even for small input changes.
+
+To have better performance (and to support non-invertible aggregations, e.g., min/max), we need to extend what the `broadcast` part of prefix sum is doing (`aggregate` can stay the same):
+  - `queries` will contain intervals specified by two indexes.
+  - `requests`: We can similarly compute a set of requests from `queries`. The change will only be inside the `flat_map`.
+  - `full_ranges`, `zero_ranges`, `used_ranges` stay the same.
+  - `init_states` won’t start at position 0, but at the lower end of the intervals in `queries`
+  - The iteration at the end will be mostly the same.
+
+#### 3. FIRST_VALUE / LAST_VALUE / NTH_VALUE
+
+For example: For each city, compute how many times it is smaller than the biggest city in the same state:
+```sql
+SELECT state, name, pop,
+       CAST(FIRST_VALUE(pop)
+              OVER (PARTITION BY state ORDER BY pop DESC)
+            AS float
+       ) / pop
+FROM cities;
+```
+
+These also operate based on a **frame**, similarly to window aggregations. (The above example query doesn't specify a frame, therefore it uses the default frame: from the beginning of the partition to the current row) They can be similarly implemented to window aggregations, i.e., we could “sum” up the relevant interval (that is not necessarily a prefix) with an appropriate sum function.
+
+Alternatively, we could make these a bit faster (except for NTH_VALUE) if we just find the index of the relevant end of the interval (i.e., left end for FIRST_VALUE), and then self-join.
+
+(And there are some special cases when we can transform away the window function usage: FIRST_VALUE with UNBOUNDED PRECEDING and LAST_VALUE with UNBOUNDED FOLLOWING should be transformed to just a (non-windowed) grouped aggregation + self-join instead of prefix sum trickery. Also, similarly for the case when there is no ORDER BY.)
```

Comment:
This sounds like a paragraph, not a parenthesis. :-)

Reply:
I removed the parenthesis, and extended the text a bit :)